packages/appium/lib/config-file.js

```diff
@@ -0,0 +1,293 @@
+// @ts-check
+
+import log from './logger';
+import _ from 'lodash';
+import {lilconfig} from 'lilconfig';
+import yaml from 'yaml';
+import Ajv from 'ajv';
+import addFormats from 'ajv-formats';
+import {fs} from '@appium/support';
+import betterAjvErrors from '@sidvind/better-ajv-errors';
+import schema from './appium.schema.json';
+
+/**
+ * lilconfig loader to handle `.yaml` files
+ * @type {import('lilconfig').LoaderSync}
+ */
+function yamlLoader (_, content) {
+  return yaml.parse(content);
+}
+
+/**
+ * Reads an Appium config file; searches for one if no `filepath` specified.
+ * @param {string} [filepath] - Explicit path to config file
+ * @returns {Promise<ReadConfigFileResult>}
+ */
+async function findConfigFile (filepath) {
+  const lc = lilconfig('appium', {
+    loaders: {
+      '.yaml': yamlLoader,
+      '.yml': yamlLoader,
+    },
+  });
+  if (filepath) {
+    log.debug(`Attempting to load config at filepath ${filepath}`);
+    try {
+      // removing "await" will cause any rejection to _not_ be caught in this block!
+      return await lc.load(filepath);
+    } catch (err) {
+      if (err.code === 'ENOENT') {
+        err.message = `Config file not found at user-provided path: ${filepath}`;
+      } else if (err instanceof SyntaxError) {
+        err.message = `Config file at user-provided path ${filepath} is invalid:\n${err.message}`;
+      }
+      throw err;
+    }
+  }
+  log.debug('No config file specified; searching...');
+  const result = await lc.search();
+  if (!result || !result.filepath) {
+    log.debug('Could not find a config file');
+  }
+  return result;
+}
+
+/**
+ * Inits Ajv, compiles the schema, and returns the validation function.
+ * Whenever this function is called, the validation function's `errors` property will be reset.
+ * If there are errors during validation, it will be set again.
+ */
+const validator = _.once(
+  /**
+   * @returns {import('ajv').ValidateFunction<import('./appium-config').AppiumConfigurationSchema>}
+   */
+  () => {
```

Comment:
I prefer to use named functions in such context, since their names are visible in the debugger stack